<?xml version="1.0"?>
<launch>

  <!-- This launch file provides Externally Gudied Motion (EGM)-based and Robot Web Services (RWS)-based
       connections to ABB robots. Requires the ABB RobotWare option Externally Gudied Motion [689-1].
         ** a new parameter is added to control J23-linkage correction **

       One node is started, to supply both low-level robot communication
       and higher-level actionlib support:
         - joint_trajectory_action : actionlib interface to control robot motion,
                                     publishes current joint positions and robot state data and
                                     command robot motion by sending motion points to robot

    Usage:
      robot_interface.launch robot_ip:=<value> [robot_web_services_port:=80] [J23_coupled:=false] [dual_armed:=false]
  -->
  
  <!-- robot_ip: IP-address of the robot -->
  <arg name="robot_ip" />
  
  <!-- rws_port: Port number for the RWS HTTP server's port -->
  <arg name="robot_web_services_port" default="80" />
  
  <!-- J23_coupled: set TRUE to apply correction for J2/J3 parallel linkage (not used at the moment) -->
  <arg name="J23_coupled" default="false" />
  
  <!-- dual_armed: set TRUE to indicate the use of a dual ABB armed robot (e.g. IRB14000) -->
  <arg name="dual_armed" default="false" />
  
  <!-- copy the specified arguments to the Parameter Server, for use by the node below -->
  <param name="joint_trajectory_action/robot_ip_address" type="str" value="$(arg robot_ip)"/>
  <param name="joint_trajectory_action/robot_web_services_port" type="str" value="$(arg robot_web_services_port)"/>
  <param name="joint_trajectory_action/J23_coupled" type="bool" value="$(arg J23_coupled)"/>
  <param name="joint_trajectory_action/dual_armed" type="bool" value="$(arg dual_armed)"/>
  
  <!-- increase the allowed execution duration a little bit -->
  <param name="move_group/trajectory_execution/allowed_execution_duration_scaling" value="4"/>
  
  <!-- joint_trajectory_action: provides: actionlib interface for high-level robot control
                                          publishes joint positions and robot-state data 
                                          sends robot motion commands by STREAMING path to robot
                                          (using EGM connection to robot) -->
  <node pkg="abb_egm_driver" type="joint_trajectory_action" name="joint_trajectory_action"/>
  
</launch>
